casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -281,32 +283,47 @@ class Running[F[_]: Concurrent: BlockStore: CasperBufferStorage: BlockRetriever:
   /**
     * Message that relates to a block A (e.g. block message or block hash broadcast message) shall be considered
     * as repeated and ignored if block A is
-    * 1. marked as received in BlocksRetriever and pending adding to CasperBuffer or
-    * 2. added do CasperBuffer and pending adding to DAG or
-    * 3. already added to DAG.
+    * 1. BlockIsWaitingForCasper  -> ready to be replayed but Casper is busy, so processing is postponed
+    * 2. BlockIsInProcessing      -> currently replayed by Casper
+    * 3. BlockIsInCasperBuffer    -> is missing dependencies so added to CasperBuffer
+    *                                and waiting for all dependencies to be filled
+    * 4. BlockIsInDag             -> is already added to the DAG.
+    * 5. BlockIsReceived          -> is received, but BlockRetriever did not process it yet (rare case).
     * Otherwise - that is a new message and shall be processed normal way.
     *
     * @param hash Block hash
     * @return If message should be ignored
     */
   private def ignoreCasperMessage(hash: BlockHash): F[IgnoreCasperMessageStatus] =
-    BlockRetriever[F]
-      .received(hash)
-      .ifM(
-        IgnoreCasperMessageStatus(doIgnore = true, BlockIsWaitingForCasper).pure[F],
-        CasperBufferStorage[F]
-          .contains(hash)
-          .ifM(
-            IgnoreCasperMessageStatus(doIgnore = true, BlockIsInCasperBuffer).pure[F],
-            casper.blockDag.flatMap(
-              _.contains(hash)
+    for {
+      received <- BlockRetriever[F].received(hash)
+      r <- BlockRetriever[F].getEnqueuedToCasper
+            .map(_.contains(hash))
+            .ifM(
+              IgnoreCasperMessageStatus(doIgnore = true, BlockIsWaitingForCasper).pure[F],
+              casper.getBlocksInProcessing
+                .map(_.contains(hash))
                 .ifM(
-                  IgnoreCasperMessageStatus(doIgnore = true, BlockIsInDag).pure[F],
-                  IgnoreCasperMessageStatus(doIgnore = false, DoNotIgnore).pure[F]
+                  IgnoreCasperMessageStatus(doIgnore = true, BlockIsInProcessing).pure[F],
+                  CasperBufferStorage[F]
+                    .contains(hash)
+                    .ifM(
+                      IgnoreCasperMessageStatus(doIgnore = true, BlockIsInCasperBuffer).pure[F],
+                      casper.blockDag.flatMap(
+                        _.contains(hash)
+                          .ifM(
+                            IgnoreCasperMessageStatus(doIgnore = true, BlockIsInDag).pure[F],
+                            if (received)
+                              IgnoreCasperMessageStatus(doIgnore = true, BlockIsReceived).pure[F]
+                            else
+                              IgnoreCasperMessageStatus(doIgnore = false, DoNotIgnore).pure[F]
```

Comment:
This is little scary.

Reply:
@tgrospic do these comment help? 
https://github.com/rchain/rchain/blob/f01e1b9bb798b4061195c9dd1f45cdfc806ae0ae/casper/src/main/scala/coop/rchain/casper/engine/Running.scala#L299-L330

Reply:
Yes, thanks. I just wanted to say, it's not easy to follow all these if's in the head.